go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,306 @@
+
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+// holds the source of truth for EgressFirewall DNS names on a node
+type EgressDNS struct {
+	// Protects pdMap/namespaces operations
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// name of the node
+	nodeName string
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows to get object using the informer cache
+	wf factory.NodeWatchFactory
+	k  kube.Interface
+
+	added          chan string
+	addFinished    chan struct{}
+	removed        chan string
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	namespaces  map[string]struct{} //the namespaces that this DNSName applies to
+}
+
+func NewEgressDNS(nodeName string, watchFactory factory.NodeWatchFactory, k kube.Interface, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:        dnsInfo,
+		nodeName:   nodeName,
+		dnsEntries: make(map[string]*dnsEntry),
+		wf:         watchFactory,
+		k:          k,
+
+		added:          make(chan string, 1),
+		addFinished:    make(chan struct{}, 1),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(dnsNames []string, namespace string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	for _, dnsName := range dnsNames {
+		if _, exists := e.dnsEntries[dnsName]; !exists {
+			e.dnsEntries[dnsName] = &dnsEntry{
+				namespaces: map[string]struct{}{
+					namespace: {},
+				},
+			}
+			e.signalAdded(dnsName)
+		} else {
+			// only need to add the namespace to the struct
+			e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+		}
+	}
+	return nil
+
+}
+
+//updateEntryForName takes a dnsName and creates/updates the EgressDNS's
+// dnsEntries map for the given dnsName assumes that you have locked the EgressDNS object
+// before calling this function
+func (e *EgressDNS) updateEntryForName(dnsName string) error {
+	ips := e.dns.GetIPs(dnsName)
+	//TODO: if the old and new resolutions are the same should not update the nodeDNSInfo
+	e.dnsEntries[dnsName].dnsResolves = ips
+
+	resultErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		//update the nodeDNSInfo object so the master can coordinate all the nodes ip addresses
+		// if the node has not reported any DNS addresses before create the nodeDNSObject
+		nodeDNSInfo, err := e.wf.GetNodeDNSInfo(e.nodeName)
+		existed := true
+		if errors.IsNotFound(err) {
+			// the object *MOST LIKELY* has not been created yet, so start making it
+			nodeDNSInfo = &nodednsinfoapi.NodeDNSInfo{
```

Comment:
You should set the OwnerReference to the Node object here, so it will be automatically cleaned up if the node goes away.

Reply:
I see what the OwnerReference is but is there a proper way to set it?

Reply:
There are a couple of examples to crib from - like https://github.com/openshift/cluster-network-operator/blob/446f2c5c35faa1bc7ab91d2deee16cbf239710aa/pkg/controller/operconfig/operconfig_controller.go#L294 - @danielmellado also went down this trip recently.